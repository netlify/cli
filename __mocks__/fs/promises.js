// eslint-disable-next-line no-undef, @typescript-eslint/no-require-imports
const { fs } = require('memfs')

// eslint-disable-next-line no-undef
module.exports = fs.promises
