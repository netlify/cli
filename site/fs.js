'use strict'

const fs = require('fs').promises
const path = require('path')
const util = require('util')

const rimraf = require('rimraf')

const rimrafAsync = util.promisify(rimraf)

const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch (_) {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = path.join(src, item)
      const destPath = path.join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(path.dirname(filePath), { recursive: true })
  } catch (_) {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}

module.exports = {
  copyDirRecursiveAsync,
  ensureFilePathAsync,
  removeRecursiveAsync,
}
