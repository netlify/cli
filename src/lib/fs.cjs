// @ts-check
const {
  constants,
  promises: { access, readFile, rm, stat },
} = require('fs')
const { version } = require('process')

const del = require('del')
const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */
const readFileAsyncCatchError = async (filePath) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */
const rmdirRecursiveAsync = async (path) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')

module.exports = {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
