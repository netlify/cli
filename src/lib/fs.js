const fs = require('fs')
const pathType = require('path-type')
const { promisify } = require('util')

const statAsync = promisify(fs.stat)
const readFileAsync = promisify(fs.readFile)
const writeFileAsync = promisify(fs.writeFile)
const accessAsync = promisify(fs.access)

const readFileAsyncCatchError = async filepath => {
  try {
    return { content: await readFileAsync(filepath) }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async filePath => {
  try {
    await accessAsync(filePath, fs.F_OK)
    return true
  } catch (_) {
    return false
  }
}

const isFileAsync = filePath => pathType.isFile(filePath)

module.exports = { statAsync, readFileAsync, readFileAsyncCatchError, writeFileAsync, fileExistsAsync, isFileAsync }
