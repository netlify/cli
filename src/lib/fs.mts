// @ts-check
const {
  
  constants,
  
  promises: { access, readFile, rm, stat },
} = require('fs')

const { version } = require('process')

import del from 'del'

const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */

const readFileAsyncCatchError = async (filePath: $TSFixMe) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}


const fileExistsAsync = async (filePath: $TSFixMe) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */

const rmdirRecursiveAsync = async (path: $TSFixMe) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */

const isType = async (filePath: $TSFixMe, type: $TSFixMe) => {
  try {
    const stats = await stat(filePath)
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    
    if ((error_ as $TSFixMe).code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */

const isFileAsync = (filePath: $TSFixMe) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */

const isDirectoryAsync = (filePath: $TSFixMe) => isType(filePath, 'isDirectory')

export default {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
