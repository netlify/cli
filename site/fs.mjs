import { promises as fs } from 'fs'
import { join, dirname } from 'path'
import { promisify } from 'util'

import rimraf from 'rimraf'

const rimrafAsync = promisify(rimraf)

export const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = join(src, item)
      const destPath = join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

export const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(dirname(filePath), { recursive: true })
  } catch {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

export const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}
