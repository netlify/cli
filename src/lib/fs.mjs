// @ts-check
import { constants } from 'fs'
import { access, stat } from 'fs/promises'

export const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
export const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
export const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')
