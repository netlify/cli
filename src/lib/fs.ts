import { constants } from 'fs'
import { access, stat } from 'fs/promises'

const isErrnoException = (value: unknown): value is NodeJS.ErrnoException =>
  value instanceof Error && Object.hasOwn(value, 'code')

export const fileExistsAsync = async (filePath: string) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 */
const isType = async (filePath: string, type: 'isFile' | 'isDirectory') => {
  try {
    const stats = await stat(filePath)
    if (type === 'isFile') return stats.isFile()
    return stats.isDirectory()
  } catch (error) {
    if (isErrnoException(error) && error.code === 'ENOENT') {
      return false
    }

    throw error
  }
}

/**
 * Checks if the provided filePath is a file
 */
export const isFileAsync = async (filePath: string) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 */
export const isDirectoryAsync = async (filePath: string) => isType(filePath, 'isDirectory')
