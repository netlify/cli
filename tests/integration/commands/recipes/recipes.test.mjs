import { readFile } from 'fs/promises'
import { resolve } from 'path'

import { parse } from 'comment-json'
import execa from 'execa'
import { describe, test } from 'vitest'

import { callCli } from '../../utils/call-cli.mjs'
import { cliPath } from '../../utils/cli-path.mjs'
import { CONFIRM, NO, answerWithValue, handleQuestions } from '../../utils/handle-questions.mjs'
import { withSiteBuilder } from '../../utils/site-builder.mjs'
import { normalize } from '../../utils/snapshots.mjs'

describe.concurrent('commands/recipes', () => {
  test('Shows a list of all the available recipes', async (t) => {
    const cliResponse = await callCli(['recipes:list'])

    t.expect(normalize(cliResponse)).toMatchSnapshot()
  })

  test('Generates a new VS Code settings file if one does not exist', async (t) => {
    await withSiteBuilder('repo', async (builder) => {
      await builder.buildAsync()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `A new VS Code settings file will be created at ${settingsPath}`,
          answer: CONFIRM,
        },
      ])

      await childProcess

      const settings = JSON.parse(await readFile(`${builder.directory}/.vscode/settings.json`))

      t.expect(settings['deno.enable']).toBe(true)
      t.expect(settings['deno.importMap']).toEqual('.netlify/edge-functions-import-map.json')
      t.expect(settings['deno.enablePaths']).toStrictEqual(['netlify/edge-functions'])
    })
  })

  test('Updates an existing VS Code settings file', async (t) => {
    await withSiteBuilder('repo', async (builder) => {
      await builder
        .withContentFile({
          path: '.vscode/settings.json',
          content: JSON.stringify({ 'deno.enablePaths': ['/some/path'], someSetting: 'value' }),
        })
        .buildAsync()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `There is a VS Code settings file at ${settingsPath}. Can we update it?`,
          answer: CONFIRM,
        },
      ])

      await childProcess

      const settings = JSON.parse(await readFile(`${builder.directory}/.vscode/settings.json`))

      t.expect(settings.someSetting).toEqual('value')
      t.expect(settings['deno.enable']).toBe(true)
      t.expect(settings['deno.importMap']).toEqual('.netlify/edge-functions-import-map.json')
      t.expect(settings['deno.enablePaths']).toStrictEqual(['/some/path', 'netlify/edge-functions'])
    })
  })

  test('Does not generate a new VS Code settings file if the user does not confirm the prompt', async (t) => {
    await withSiteBuilder('repo', async (builder) => {
      await builder.buildAsync()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `A new VS Code settings file will be created at ${settingsPath}`,
          answer: answerWithValue(NO),
        },
      ])

      await childProcess

      const error = await readFile(`${builder.directory}/.vscode/settings.json`).catch((error_) => error_)
      t.expect(error.code).toEqual('ENOENT')
    })
  })

  test('Handles JSON with comments', async (t) => {
    await withSiteBuilder('repo', async (builder) => {
      const comment = '// sets value for someSetting'
      await builder
        .withContentFile({
          path: '.vscode/settings.json',
          content: `{
          "deno.enablePaths":["/some/path"],
          "someSetting":"value" ${comment}
        }`,
        })
        .buildAsync()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `There is a VS Code settings file at ${settingsPath}. Can we update it?`,
          answer: CONFIRM,
        },
        {
          question: 'The Deno VS Code extension is recommended. Would you like to install it now?',
          answer: answerWithValue(NO),
        },
      ])

      await childProcess

      const settingsText = await readFile(`${builder.directory}/.vscode/settings.json`, { encoding: 'utf8' })
      t.expect(settingsText.includes(comment)).toBe(true)

      const settings = parse(settingsText, null, true)
      t.expect(settings.someSetting).toEqual('value')
      t.expect(settings['deno.enable']).toBe(true)
      t.expect(settings['deno.importMap']).toEqual('.netlify/edge-functions-import-map.json')
      t.expect([...settings['deno.enablePaths']]).toStrictEqual(['/some/path', 'netlify/edge-functions'])
    })
  })

  test('Suggests closest matching recipe on typo', async (t) => {
    const cliResponse = await callCli(['recipes', 'vsc'])

    t.expect(normalize(cliResponse)).toMatchSnapshot()
  })
})
