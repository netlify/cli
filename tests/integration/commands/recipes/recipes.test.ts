import fs from 'node:fs/promises'
import path from 'node:path'

import { parse } from 'comment-json'
import execa from 'execa'
import { describe, test } from 'vitest'

import { callCli } from '../../utils/call-cli.js'
import { cliPath } from '../../utils/cli-path.js'
import { CONFIRM, NO, answerWithValue, handleQuestions } from '../../utils/handle-questions.js'
import { withSiteBuilder } from '../../utils/site-builder.js'
import { normalize } from '../../utils/snapshots.js'

describe.concurrent('commands/recipes', () => {
  test('Shows a list of all the available recipes', async (t) => {
    const cliResponse = (await callCli(['recipes:list'])) as string

    t.expect(normalize(cliResponse)).toMatchSnapshot()
  })

  test('Generates a new VS Code settings file if one does not exist', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      await builder.build()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = path.resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `A new VS Code settings file will be created at ${settingsPath}`,
          answer: CONFIRM,
        },
      ])

      await childProcess

      const settings = JSON.parse(
        await fs.readFile(path.join(builder.directory, `.vscode/settings.json`), 'utf8'),
      ) as unknown

      t.expect(settings).toHaveProperty('deno.enable', true)
      t.expect(settings).toHaveProperty('deno.importMap', '.netlify/edge-functions-import-map.json')
      t.expect(settings).toHaveProperty('deno.enablePaths', ['netlify/edge-functions'])
    })
  })

  test('Updates an existing VS Code settings file', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      await builder
        .withContentFile({
          path: '.vscode/settings.json',
          content: JSON.stringify({ 'deno.enablePaths': ['/some/path'], someSetting: 'value' }),
        })
        .build()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = path.resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `There is a VS Code settings file at ${settingsPath}. Can we update it?`,
          answer: CONFIRM,
        },
      ])

      await childProcess

      const settings = JSON.parse(
        await fs.readFile(path.join(builder.directory, `.vscode/settings.json`), 'utf8'),
      ) as unknown

      t.expect(settings).toHaveProperty('someSetting', 'value')
      t.expect(settings).toHaveProperty('deno.enable', true)
      t.expect(settings).toHaveProperty('deno.importMap', '.netlify/edge-functions-import-map.json')
      t.expect(settings).toHaveProperty('deno.enablePaths', ['/some/path', 'netlify/edge-functions'])
    })
  })

  test('Does not generate a new VS Code settings file if the user does not confirm the prompt', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      await builder.build()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = path.resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `A new VS Code settings file will be created at ${settingsPath}`,
          answer: answerWithValue(NO),
        },
      ])

      await childProcess

      try {
        await fs.readFile(path.join(builder.directory, `.vscode/settings.json`), 'utf8')
        t.expect.unreachable()
      } catch (err) {
        t.expect(err).toHaveProperty('code', 'ENOENT')
      }
    })
  })

  test('Handles JSON with comments', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      const comment = '// sets value for someSetting'
      await builder
        .withContentFile({
          path: '.vscode/settings.json',
          content: `{
          "deno.enablePaths":["/some/path"],
          "someSetting":"value" ${comment}
        }`,
        })
        .build()

      const childProcess = execa(cliPath, ['recipes', 'vscode'], {
        cwd: builder.directory,
      })
      const settingsPath = path.resolve(builder.directory, '.vscode', 'settings.json')

      handleQuestions(childProcess, [
        {
          question: `There is a VS Code settings file at ${settingsPath}. Can we update it?`,
          answer: CONFIRM,
        },
        {
          question: 'The Deno VS Code extension is recommended. Would you like to install it now?',
          answer: answerWithValue(NO),
        },
      ])

      await childProcess

      const settingsText = await fs.readFile(path.join(builder.directory, `.vscode/settings.json`), 'utf8')
      t.expect(settingsText.includes(comment)).toBe(true)

      const settings = parse(settingsText, null, true)
      t.expect(settings).toHaveProperty('someSetting', 'value')
      t.expect(settings).toHaveProperty('deno.enable', true)
      t.expect(settings).toHaveProperty('deno.importMap', '.netlify/edge-functions-import-map.json')
      t.expect(settings).toHaveProperty('deno.enablePaths', ['/some/path', 'netlify/edge-functions'])
    })
  })

  test('Suggests closest matching recipe on typo', async (t) => {
    const cliResponse = (await callCli(['recipes', 'vsc'])) as string

    t.expect(normalize(cliResponse)).toMatchSnapshot()
  })
})
