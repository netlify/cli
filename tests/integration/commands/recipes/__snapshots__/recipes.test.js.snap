// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`commands/recipes > Shows a list of all the available recipes 1`] = `
".-----------------------------------------------------------------------------------------.
|                              Usage: netlify recipes <name>                              |
|-----------------------------------------------------------------------------------------|
|     Name      |                               Description                               |
|---------------|-------------------------------------------------------------------------|
| blobs-migrate | Migrate legacy Netlify Blobs stores                                     |
| vscode        | Create VS Code settings for an optimal experience with Netlify projects |
'-----------------------------------------------------------------------------------------'"
`;

exports[`commands/recipes > Suggests closest matching recipe on typo 1`] = `
"â—ˆ vsc is not a valid recipe name.
? Did you mean vscode (y/N) [28D[28C"
`;
